node/src/main/scala/org/bitcoins/node/networking/peer/DataMessageHandler.scala

```diff
@@ -37,17 +56,93 @@ class DataMessageHandler(appConfig: ChainAppConfig)(
           val lastHash = headers.last.hash
           peerMsgSender.sendGetHeadersMessage(lastHash)
         }
+      case msg: BlockMessage =>
+        callbacks.onBlockReceived(msg.block)
+        FutureUtil.unit
+      case msg: TransactionMessage =>
+        callbacks.onTxReceived(msg.transaction)
+        FutureUtil.unit
+      case msg: MerkleBlockMessage =>
+        callbacks.onMerkleBlockReceived(msg.merkleBlock)
+        FutureUtil.unit
       case invMsg: InventoryMessage =>
         handleInventoryMsg(invMsg = invMsg, peerMsgSender = peerMsgSender)
     }
   }
 
+  /**
+    * We send `getdata` requests for all filtered blocks.
+    * That means we request a `merkleblock` message, given we
+    * have configured a bloom filter with this peer prior to this
+    */
   private def handleInventoryMsg(
       invMsg: InventoryMessage,
       peerMsgSender: PeerMessageSender): Future[Unit] = {
-    logger.info(s"Received inv=${invMsg}")
+    logger.info(
+      s"Received inv message. Inventory count: ${invMsg.inventoryCount.toInt}")
+    logger.info(
+      s"First 5 inventories: ${invMsg.inventories.take(5).mkString(", ")}")
+
+    val inventories: Seq[Inventory] = {
+      val listOfOpts = for {
+        inventory <- invMsg.inventories
+      } yield {
+        inventory.typeIdentifier match {
+          // changing the inventory type might be naughty,
+          // but we're not interested
+          // in the complete block
+          case MsgBlock =>
+            Some(
+              Inventory(TypeIdentifier.MsgFilteredBlock, hash = inventory.hash))
+          case MsgFilteredBlock => Some(inventory)
+          case MsgTx            => Some(inventory)
+          case _: MsgUnassigned => None
+        }
+      }
+      listOfOpts.flatten
+    }
+
+    if (inventories.isEmpty) {
+      logger.info(
+        s"Found no inventories of interest to us, not sending getdata message")
+    } else {
+      val getdata = GetDataMessage(inventories)
+      logger.info(s"Sending getdata message for ${inventories.mkString(", ")}")
+      peerMsgSender.sendMsg(getdata)
+    }
 
     FutureUtil.unit
 
   }
 }
+
+object DataMessageHandler extends BitcoinSLogger {
+
+  /** Callback for handling a received block */
+  type OnBlockReceived = Block => Unit
```

Comment:
I have a feeling that these signatures should be returning `Future[Unit]` so we don't block on whatever they are doing in this function?

Reply:
I'm not sure that's needed. Consider this:
```scala
def process(func: () => Any) = {
  println("before in process")
  func()
  println("after in process")
} 

val futureFunc: () => Future[Unit] = { () => 
  println("before the future")
  Future { 
    Thread.sleep(3000)
    println("Just slept")
  }
} 
```

Calling `process(futureFunc)` gives the following output: 
```scala
@ process(futureFunc) 
before in process
before the future
after in process
Just slept
```

So it's really up to the user at call site to block or not to block. But that's always the case, we can't prevent people from blocking if they really want to.